moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -874,83 +877,193 @@ bool TimeOptimalTrajectoryGeneration::computeTimeStamps(robot_trajectory::RobotT
-  }
-  else
-  {
-    ROS_WARN_NAMED(LOGNAME, "Invalid max_velocity_scaling_factor %f specified, defaulting to %f instead.",
-                   max_velocity_scaling_factor, velocity_scaling_factor);
-  }
+  double velocity_scaling_factor = verifyScalingFactor(max_velocity_scaling_factor);
+  double acceleration_scaling_factor = verifyScalingFactor(max_acceleration_scaling_factor);
 
-  double acceleration_scaling_factor = 1.0;
-  if (max_acceleration_scaling_factor > 0.0 && max_acceleration_scaling_factor <= 1.0)
+  // Get the velocity and acceleration limits for all active joints
+  const moveit::core::RobotModel& rmodel = group->getParentModel();
+  const std::vector<std::string>& vars = group->getVariableNames();
+  std::vector<size_t> active_joint_indices;
+  if (!group->computeJointVariableIndices(group->getActiveJointModelNames(), active_joint_indices))
   {
-    acceleration_scaling_factor = max_acceleration_scaling_factor;
+    ROS_ERROR_NAMED(LOGNAME, "Failed to get active variable indices.");
   }
-  else if (max_acceleration_scaling_factor == 0.0)
+
+  const size_t num_active_joints = active_joint_indices.size();
+  Eigen::VectorXd max_velocity(num_active_joints);
+  Eigen::VectorXd max_acceleration(num_active_joints);
+  for (size_t idx = 0; idx < num_active_joints; ++idx)
   {
-    ROS_DEBUG_NAMED(LOGNAME, "A max_acceleration_scaling_factor of 0.0 was specified, defaulting to %f instead.",
-                    acceleration_scaling_factor);
+    // For active joints only (skip mimic joints and other types)
+    const moveit::core::VariableBounds& bounds = rmodel.getVariableBounds(vars[active_joint_indices[idx]]);
+
+    // Limits need to be non-zero, otherwise we never exit
+    if (bounds.velocity_bounded_)
+    {
+      if (bounds.max_velocity_ <= 0.0)
+      {
+        ROS_ERROR_NAMED(LOGNAME, "Invalid max_velocity %f specified for '%s', must be greater than 0.0",
+                        bounds.max_velocity_, vars[idx].c_str());
+        return false;
+      }
+      max_velocity[idx] =
+          std::min(std::fabs(bounds.max_velocity_), std::fabs(bounds.min_velocity_)) * velocity_scaling_factor;
+    }
+    else
+    {
+      max_velocity[idx] = DEFAULT_VELOCITY_LIMIT;
+    }
+
+    if (bounds.acceleration_bounded_)
+    {
+      if (bounds.max_acceleration_ < 0.0)
+      {
+        ROS_ERROR_NAMED(LOGNAME, "Invalid max_acceleration %f specified for '%s', must be greater than 0.0",
+                        bounds.max_acceleration_, vars[idx].c_str());
+        return false;
+      }
+      max_acceleration[idx] = std::min(std::fabs(bounds.max_acceleration_), std::fabs(bounds.min_acceleration_)) *
+                              acceleration_scaling_factor;
+    }
+    else
+    {
+      max_acceleration[idx] = DEFAULT_ACCELERATION_LIMIT;
+    }
   }
-  else
+
+  return doTimeParameterizationCalculations(trajectory, max_velocity, max_acceleration);
+}
+
+bool TimeOptimalTrajectoryGeneration::computeTimeStamps(
```

Comment:
This method has strong overlap with the other computeTimeStamps method.
Better factor out common code into a separate helper function!

Reply:
Completely agree with you but won't address now. It's not really related to this "torque-limited" feature.

BTW I'm no longer at PickNik. I'm working on ag robots with AppliedLogix now.

Reply:
You seem to like programming by copy-and-paste! I'm trying to cleanup the redundancy and discovered some more issues in this PR, namely you are incorrectly indexing the joints. Could you please open your PR #3426 for additional commits?